### DISCUSSION TITLE
Criteria API（NativeSql）を使用した複数エンティティへの射影

### DISCUSSION THREAD

**Discussion by ironclad-15256304** [domaframework/doma#974] [Q&A] [answered]
### 相談内容

いつも大変お世話になっております。

私の参加しているプロジェクトでは、「[複数エンティティの選択 (NativeSql)](https://doma.readthedocs.io/en/latest/criteria-api/#multiple-entity-selection-nativesql)」を使用し、一つのクエリで複数のエンティティを一括で取得したいシチュエーションが頻繁に登場します。
そのうち、大量の情報を一括で取得するようなクエリでは、目的に応じてその処理で必要な情報が限られていることが多いです。そのため、「[射影(Entityql, NativeSql)](https://doma.readthedocs.io/en/latest/criteria-api/#column-projection-and-mapping-entityql-nativesql)」のような形で、取得列そのものに制限をかけることでデータ量を抑える実装を可能であれば行いたいと考えております。

このような場合、NativeSqlで実現できるような実装方法はございますでしょうか。もしくは、事前にその目的に応じてフィールドの数を減らしたEntityを別で定義しておいたりなど、別の形で実現した方が良いでしょうか。
また、似た要件で、Entityqlのassociateで関連entityまで取得したい場合にも同様に、Entity毎に取得列を指定することは出来ますでしょうか。
お忙しいところ大変恐縮ですが、アイデアや実現方法等ご教示頂けますと幸いです。

### 動作環境

+ Java Framework: Spring boot 3.1.0
+ Doma: 2.53.1（doma-spring-boot 1.7.0）
+ DB: postres もしくは oracle

**Comment by nakamura-to** [comment, marked as answer]
下記のように一度RowのListで受け取り、1行ごとにRowを任意のEntity群に詰め替えてはいかがでしょうか？

```java
Employee_ e = new Employee_();
Address_ a = new Address_();

List<Row> rows = nativeSql.from(e).leftJoin(a, ...).selectAsRow(e.employeeName, e.employeeNo, a.street).fetch();
List<Tuple2<Employee, Address>> list = rows.stream().map(row -> {...}); // EmployeeやAddressのエンティティに詰め替える
```

ドキュメントとしては下記を参照ください。
https://doma.readthedocs.io/en/latest/criteria-api/#column-projection-nativesql

**Reply by ironclad-15256304** [reply]
たしかに、NativeSqlを使用する場合、一度にEntityで受け取ることにこだわらず自分で詰め直しても何も問題ないですね…
参考になります。有難うございます。

**Comment by nakamura-to** [comment]
> また、似た要件で、Entityqlのassociateで関連entityまで取得したい場合にも同様に、Entity毎に取得列を指定することは出来ますでしょうか。

できないです。

> 事前にその目的に応じてフィールドの数を減らしたEntityを別で定義しておいたりなど、別の形で実現した方が良いでしょうか。

この方法は有効だと思います。

**Reply by ironclad-15256304** [reply]
承知しました。ご回答いただき有難うございます。
